python/tskit/trees.py

```diff
@@ -9291,25 +9290,156 @@ def ibd_segments(
+        sequence_length = self.sequence_length
+
+        nodes_parent = np.full(num_nodes, tskit.NULL)
+        sample_counts = np.zeros((num_nodes, num_sample_sets))
+        coalescing_pairs = np.zeros((num_windows, num_nodes, num_indexes))
+
+        for i, s in enumerate(sample_sets):
+            sample_counts[s, i] = 1
+
+        position = 0.0
+        w, a, b = 0, 0, 0
+        while position < sequence_length:
+            remainder = sequence_length - position
+
+            while b < num_edges and remove_position[b] == position:  # edges out
+                e = remove_index[b]
+                p = edges_parent[e]
+                c = edges_child[e]
+                nodes_parent[c] = tskit.NULL
+                inside = sample_counts[c]
+                while p != tskit.NULL:
+                    outside = sample_counts[p] - sample_counts[c]
+                    for i, (j, k) in enumerate(indexes):
+                        weight = inside[j] * outside[k] + inside[k] * outside[j]
+                        coalescing_pairs[w, p, i] -= weight * remainder
+                    c, p = p, nodes_parent[p]
+                p = edges_parent[e]
+                while p != tskit.NULL:
+                    sample_counts[p] -= inside
+                    p = nodes_parent[p]
+                b += 1
+
+            while a < num_edges and insert_position[a] == position:  # edges in
+                e = insert_index[a]
+                p = edges_parent[e]
+                c = edges_child[e]
+                nodes_parent[c] = p
+                inside = sample_counts[c]
+                while p != tskit.NULL:
+                    sample_counts[p] += inside
+                    p = nodes_parent[p]
+                p = edges_parent[e]
+                while p != tskit.NULL:
+                    outside = sample_counts[p] - sample_counts[c]
+                    for i, (j, k) in enumerate(indexes):
+                        weight = inside[j] * outside[k] + inside[k] * outside[j]
+                        coalescing_pairs[w, p, i] += weight * remainder
+                    c, p = p, nodes_parent[p]
+                a += 1
+
+            position = sequence_length
+            if b < num_edges:
+                position = min(position, remove_position[b])
+            if a < num_edges:
+                position = min(position, insert_position[a])
+
+            while w < num_windows and windows[w + 1] <= position:  # flush window
+                remainder = sequence_length - windows[w + 1]
+                for c, p in enumerate(nodes_parent):
+                    if p == tskit.NULL:
+                        continue
+                    inside = sample_counts[c]
+                    outside = sample_counts[p] - sample_counts[c]
+                    for i, (j, k) in enumerate(indexes):
+                        weight = inside[j] * outside[k] + inside[k] * outside[j]
+                        coalescing_pairs[w, p, i] -= weight * remainder / 2
```

Comment:
why is this divided by 2?

Reply:
it's because this accumulates weights from all `children * (children - 1)` ordered pairs of children for anode -- and instead we want to accumulate weights from all `children * (children - 1) / 2` unordered pairs. Both orderings of a pair produce the same weight, so we divide by two